### ISSUE TITLE
Peer Aborted the SSL Handshake

### ISSUE THREAD

**Issue by PurgePJ** [luvit/luvit#986]
Hello, I'm here today to report an issue that has been happening to a lot of users so many times and I don't find a solution to it.

Using coro-http module, depends on what sites (all of them are HTTPS) we receive that error above, "Peer Aborted the SSL Handshake".
As it is obvious, we switched to https module to see if this could fix it, but sadly, it didn't.

I don't bring any actual code example because it is just a simple GET request. This is the url i've been having problems with: 

> https://elosuite.com/loader_files/db.php

Seems like a luvit problem.

**Comment by creationix** [comment]
Thanks for the report.  So just to confirm, you're trying to just do a GET request as http client using coro-http?

**Comment by PurgePJ** [comment]
I am trying a GET request with coro-http and that's the error I get, using https module doesn't work either.

**Comment by squeek502** [comment]
Can confirm this is reproducible.

Using `https`:

```lua
local https = require('https')

local req = https.get('https://elosuite.com/loader_files/db.php', function (res)
  p(res)
end)
req:on("error", function(...)
  p("error", ...)
end)
```

Output: `'error' 'ssl'`

Using `coro-http`:

```lua
local http = require('coro-http')

coroutine.wrap(function()
  local res, body = http.request("GET", "https://elosuite.com/loader_files/db.php")
  assert(res.code == 200, res.code)
  assert(body ~= nil)
end)()
```

Output: `Uncaught Error: ...\coro-http.lua:74: Peer aborted the SSL handshake`

**Comment by Tigerism** [comment]
I also get this error with a lot of sites I try to send a request to, and it's a very frustrating error because there's nothing I can do to solve it.

Here's another example of a site that doesn't work with a GET request: https://verify.eryn.io/api/user/113691352327389188

**Comment by rphillips** [comment]
We are failing the handshake. I am not sure if it is server side or client side. If anyone owns a server that is exhibiting this issue, the SSL debug output server side would be very useful.

**Comment by rphillips** [comment]
It could be the `TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256` or `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256` suites of ciphers. Perhaps elliptic curve based handshakes are not working... something to try and pin down.

**Comment by zhaozg** [comment]
mark

**Comment by PurgePJ** [comment]
A fix to this would be kindly appreciated. I need to use other languages to perform https requests amd pass the received data to my luvit process. This is a bad practice but I don't know what else to do.

**Comment by PurgePJ** [comment]
The website above uses this:

![image](https://user-images.githubusercontent.com/20613820/26916281-054b53ac-4c2a-11e7-8aa4-0ddac912d1ab.png)

**Comment by rphillips** [comment]
OSX:
```
openssl s_client -tls1 -connect elosuite.com:443
CONNECTED(00000003)
59909:error:14094410:SSL routines:SSL3_READ_BYTES:sslv3 alert handshake failure:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-64.50.6/src/ssl/s3_pkt.c:1145:SSL alert number 40
59909:error:1409E0E5:SSL routines:SSL3_WRITE_BYTES:ssl handshake failure:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-64.50.6/src/ssl/s3_pkt.c:566:
```
Perhaps the website is broken.

**Comment by PurgePJ** [comment]
@rphillips The site is not broken. We can perform https requests to the same website using node.js. 
We encounter more sites like that one firing that SSL error. It is definitely luvit side.

**Comment by zhaozg** [comment]
You need set servername

```
zhaozg@FIREFORGE E:\work\product\openssl\win32\bin                                                                     
> openssl s_client -tls1_2 -connect elosuite.com:443 -servername elosuite.com                                          
WARNING: can't open config file: ..\win32/..\win32/ssl/openssl.cnf                                                     
CONNECTED(0000021C)                                                                                                    
depth=2 C = GB, ST = Greater Manchester, L = Salford, O = COMODO CA Limited, CN = COMODO ECC Certification Authority   
verify error:num=20:unable to get local issuer certificate                                                             
9836:error:1408D07B:SSL routines:ssl3_get_key_exchange:bad signature:.\ssl\s3_clnt.c:2038:                             
---                                                                                                                    
Certificate chain                                                                                                      
 0 s:/OU=Domain Control Validated/OU=PositiveSSL Multi-Domain/CN=sni193384.cloudflaressl.com                           
   i:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO ECC Domain Validation Secure Server CA 2      
 1 s:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO ECC Domain Validation Secure Server CA 2      
   i:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO ECC Certification Authority                   
 2 s:/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO ECC Certification Authority                   
   i:/C=SE/O=AddTrust AB/OU=AddTrust External TTP Network/CN=AddTrust External CA Root                                 
---                                                                                                                    
Server certificate                                                                                                     
-----BEGIN CERTIFICATE-----                                                                                            
MIIGmDCCBj+gAwIBAgIRAMzLRbYy33sBWfuBBaj+6XEwCgYIKoZIzj0EAwIwgZIx                                                       
CzAJBgNVBAYTAkdCMRswGQYDVQQIExJHcmVhdGVyIE1hbmNoZXN0ZXIxEDAOBgNV                                                       
BAcTB1NhbGZvcmQxGjAYBgNVBAoTEUNPTU9ETyBDQSBMaW1pdGVkMTgwNgYDVQQD                                                       
Ey9DT01PRE8gRUNDIERvbWFpbiBWYWxpZGF0aW9uIFNlY3VyZSBTZXJ2ZXIgQ0Eg                                                       
MjAeFw0xNzA1MjUwMDAwMDBaFw0xNzEyMDEyMzU5NTlaMGwxITAfBgNVBAsTGERv                                                       
bWFpbiBDb250cm9sIFZhbGlkYXRlZDEhMB8GA1UECxMYUG9zaXRpdmVTU0wgTXVs                                                       
dGktRG9tYWluMSQwIgYDVQQDExtzbmkxOTMzODQuY2xvdWRmbGFyZXNzbC5jb20w                                                       
WTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAT7ERE60Le2jX3Ih9wrj4Oy4Y/selCi                                                       
6Y/1KUntv0drgbof6R+ptyologSa+/snOoVm2gw2SZ/6W6tGE83WtHTTo4IEmTCC                                                       
BJUwHwYDVR0jBBgwFoAUQAlhZ/C8g3FP3hIILG/U1Ct2PZYwHQYDVR0OBBYEFD5k                                                       
tE4L0Vmb6ZIWlPaWdK5aBg/gMA4GA1UdDwEB/wQEAwIHgDAMBgNVHRMBAf8EAjAA                                                       
MB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjBPBgNVHSAESDBGMDoGCysG                                                       
AQQBsjEBAgIHMCswKQYIKwYBBQUHAgEWHWh0dHBzOi8vc2VjdXJlLmNvbW9kby5j                                                       
b20vQ1BTMAgGBmeBDAECATBWBgNVHR8ETzBNMEugSaBHhkVodHRwOi8vY3JsLmNv                                                       
bW9kb2NhNC5jb20vQ09NT0RPRUNDRG9tYWluVmFsaWRhdGlvblNlY3VyZVNlcnZl                                                       
ckNBMi5jcmwwgYgGCCsGAQUFBwEBBHwwejBRBggrBgEFBQcwAoZFaHR0cDovL2Ny                                                       
dC5jb21vZG9jYTQuY29tL0NPTU9ET0VDQ0RvbWFpblZhbGlkYXRpb25TZWN1cmVT                                                       
ZXJ2ZXJDQTIuY3J0MCUGCCsGAQUFBzABhhlodHRwOi8vb2NzcC5jb21vZG9jYTQu                                                       
Y29tMIIC4AYDVR0RBIIC1zCCAtOCG3NuaTE5MzM4NC5jbG91ZGZsYXJlc3NsLmNv                                                       
bYIXKi5hbGFuZGlzY291bnRzdG9yZS54eXqCDyouYWxrYXBvbjRpay5ydYIPKi5h                                                       
bmdlbGFsZWUueHl6gg4qLmJhc29ydHVtLmNvbYIOKi5lbG9zdWl0ZS5jb22CESou                                                       
Z2VudXNyZXRhaWwuY29tghIqLmhvdGVsdmlrbmVzaC5jb22CEioubmFtYXpkZWZ0                                                       
ZXJpLmNvbYILKi5uYXp6aWUudGuCDCoubnBjaGFybS5ydYINKi5veHlxaWFzLnh5                                                       
eoIUKi5wYXRyaW90LXN0dWRlbnQucnWCEyoucmVhbG11c2ljbGl2ZS5jb22CEiou                                                       
c3Vydml2YWxzZXRzLmNvbYIPKi50ZWxlY2ljYXMueHl6gg0qLnRlcmF2ZXJhLnBo                                                       
gg4qLnRoZW5lb3R2LmNvbYIMKi53ZXRsaWVwLmNmggsqLndpZGRpYS50a4ILKi53                                                       
b2RkdXkudGuCCyoud3VkZGlzLnRrggsqLnlpbm5lNS5nYYIVYWxhbmRpc2NvdW50                                                       
c3RvcmUueHl6gg1hbGthcG9uNGlrLnJ1gg1hbmdlbGFsZWUueHl6ggxiYXNvcnR1                                                       
bS5jb22CDGVsb3N1aXRlLmNvbYIPZ2VudXNyZXRhaWwuY29tghBob3RlbHZpa25l                                                       
c2guY29tghBuYW1hemRlZnRlcmkuY29tggluYXp6aWUudGuCCm5wY2hhcm0ucnWC                                                       
C294eXFpYXMueHl6ghJwYXRyaW90LXN0dWRlbnQucnWCEXJlYWxtdXNpY2xpdmUu                                                       
Y29tghBzdXJ2aXZhbHNldHMuY29tgg10ZWxlY2ljYXMueHl6ggt0ZXJhdmVyYS5w                                                       
aIIMdGhlbmVvdHYuY29tggp3ZXRsaWVwLmNmggl3aWRkaWEudGuCCXdvZGR1eS50                                                       
a4IJd3VkZGlzLnRrggl5aW5uZTUuZ2EwCgYIKoZIzj0EAwIDRwAwRAIgdqm/Yflz                                                       
/0v2NnpcrJNCn524r2/fJyZFnJyHZ0+cJucCIAu9qcjOYOWYxRLLtn3BAfCDAke6                                                       
Q1ZpPUZpd2gYSC+9                                                                                                       
-----END CERTIFICATE-----                                                                                              
subject=/OU=Domain Control Validated/OU=PositiveSSL Multi-Domain/CN=sni193384.cloudflaressl.com                        
issuer=/C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO ECC Domain Validation Secure Server CA 2    
---                                                                                                                    
No client certificate CA names sent                                                                                    
Peer signing digest: SHA256                                                                                            
Server Temp Key: ECDH, P-256, 256 bits                                                                                 
---                                                                                                                    
SSL handshake has read 3845 bytes and written 7 bytes                                                                  
---                                                                                                                    
New, (NONE), Cipher is (NONE)                                                                                          
Server public key is 256 bit                                                                                           
Secure Renegotiation IS supported                                                                                      
Compression: NONE                                                                                                      
Expansion: NONE                                                                                                        
No ALPN negotiated                                                                                                     
SSL-Session:                                                                                                           
    Protocol  : TLSv1.2                                                                                                
    Cipher    : 0000                                                                                                   
    Session-ID:                                                                                                        
    Session-ID-ctx:                                                                                                    
    Master-Key:                                                                                                        
    Key-Arg   : None                                                                                                   
    PSK identity: None                                                                                                 
    PSK identity hint: None                                                                                            
    SRP username: None                                                                                                 
    Start Time: 1497022580                                                                                             
    Timeout   : 7200 (sec)                                                                                             
    Verify return code: 20 (unable to get local issuer certificate)                                                    
---                                                                                                                    
```

**Comment by squeek502** [comment]
Can confirm that setting `'hostname'` fixes this in the `https` dep. Right now, [`'hostname'` only gets set if the `servername` key is set in the `options` passed to `https.request`](https://github.com/luvit/luvit/blob/master/deps/tls/common.lua#L156-L158), but that key [does not get populated automatically](https://github.com/luvit/luvit/blob/master/deps/tls/init.lua#L81). I don't know enough about `tls`/`openssl` to know how that should be handled.

The following does work, but it definitely seems like this should be handled automatically:
```
local https = require('https')
local parseUrl = require('http').parseUrl

local options = parseUrl("https://elosuite.com/loader_files/db.php")
options.servername = options.host

local req = https.get(options, function (res)
  p(res)
end)
```

**Comment by rphillips** [comment]
Thanks @zhaozg and @squeek502

**Comment by rphillips** [comment]
Will be in the 2.14.2 release.

https://github.com/luvit/luvit/pull/988

**Comment by rphillips** [comment]
Released 2.14.2
